### Author : Rahul Bhoyar

In [1]:
!pip install langchain langchain_openai faiss-cpu faiss-gpu

In [2]:
import os
openai_api_key = "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"  # Enter your OPENAI_API_KEY
os.environ["OPENAI_API_KEY"] = openai_api_key
print("OPENAI API key is set successfully :",openai_api_key)

OPENAI API key is set successfully : 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


In [3]:
from langchain.memory import ConversationBufferMemory
memory = ConversationBufferMemory(
    memory_key="chat_history",
    return_messages=True
)

### Conversation Retrieval Chain

In [4]:
from langchain_openai import ChatOpenAI
llm = ChatOpenAI(model_name = "gpt-4")
print("OpenAI model initiated.")
print("-"*200)
print(llm)

OpenAI model initiated.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
client=<openai.resources.chat.completions.Completions object at 0x7cf210bb4850> async_client=<openai.resources.chat.completions.AsyncCompletions object at 0x7cf210bb5d20> model_name='gpt-4' openai_api_key=SecretStr('**********') openai_proxy=''


In [5]:
VECTOR_DATABASE_PATH = "vectorstore/db_faiss"

In [6]:
from langchain.vectorstores import FAISS
from langchain_openai import OpenAIEmbeddings

In [7]:
embeddings = OpenAIEmbeddings()

In [11]:
vectordb = FAISS.load_local(VECTOR_DATABASE_PATH, embeddings)
print("Vector database loaded successfully.")
vectordb

Vector database loaded successfully.


In [12]:
# With specifying the number of arguments as  k = 15
retriever = vectordb.as_retriever(search_kwargs={"k": 15})

In [13]:
from langchain.chains import ConversationalRetrievalChain
retriever=vectordb.as_retriever()
qa = ConversationalRetrievalChain.from_llm(
    llm,
    retriever=retriever,
    memory=memory
)

In [14]:
question = "Give me the datasets about India?"
result = qa({"question": question})

/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:117: LangChainDeprecationWarning: The function `__call__` was deprecated in LangChain 0.1.0 and will be removed in 0.2.0. Use invoke instead.
  warn_deprecated(


In [15]:
result['answer']

'1. Dataset Name: India News Headlines Dataset\n   Dataset Number: 1904\n   Description: This dataset by Rohit Kulkarni contains twenty one years of headlines focusing on India. \n   Size: 93MB\n   Usability rating: 1.00\n   Tags: cities and urban areas, india, business, nlp, text, news\n   More info: [Link here](https://www.kaggle.com/datasets/therohk/india-headlines-news-dataset)\n   \n2. Dataset Name: Hotels in India\n   Dataset Number: 11\n   Description: This dataset by Aakash Shinde contains details about hotels in India such as place, travel sustainability, and ratings.\n   Size: 223KB\n   Usability rating: 0.82\n   Tags: travel, hotels and accommodations\n   More info: [Link here](https://www.kaggle.com/datasets/aakashshinde1507/hotels-in-india)\n   \n3. Dataset Name: COVID-19 in India\n   Dataset Number: 1735\n   Description: This dataset by SRK provides data on the Novel Corona Virus Disease 2019 in India. \n   Size: 758KB\n   Usability rating: 0.91\n   Tags: public health, h

1. Dataset Name: India News Headlines Dataset
   Dataset Number: 1904
   Description: This dataset by Rohit Kulkarni contains twenty one years of headlines focusing on India.
   Size: 93MB
   Usability rating: 1.00
   Tags: cities and urban areas, india, business, nlp, text, news
   More info: [Link here](https://www.kaggle.com/datasets/therohk/india-headlines-news-dataset)
   
2. Dataset Name: Hotels in India
   Dataset Number: 11
   Description: This dataset by Aakash Shinde contains details about hotels in India such as place, travel sustainability, and ratings.
   Size: 223KB
   Usability rating: 0.82
   Tags: travel, hotels and accommodations
   More info: [Link here](https://www.kaggle.com/datasets/aakashshinde1507/hotels-in-india)
   
3. Dataset Name: COVID-19 in India
   Dataset Number: 1735
   Description: This dataset by SRK provides data on the Novel Corona Virus Disease 2019 in India.
   Size: 758KB
   Usability rating: 0.91
   Tags: public health, health
   More info: [Link here](https://www.kaggle.com/datasets/sudalairajkumar/covid19-in-india)
   
4. Dataset Name: Education in India
   Dataset Number: 1462
   Description: This dataset by Rajanand Ilangovan provides district and state-wise primary & secondary school education data for 2015-16 in India.
   Size: 880KB
   Usability rating: 0.82
   Tags: india, education, primary and secondary schools
   More info: [Link here](https://www.kaggle.com/datasets/rajanand/education-in-india)

In [16]:
question = "Can you tell me which of these datasets are most important"
result = qa({"question": question})
output = result['answer']
output

'The system does not provide a ranking or importance level for these datasets. The importance of a dataset can depend on the specific needs or interests of the user. For example, if a user is interested in public health, they might find the "COVID-19 in India" dataset most important. If a user is interested in news or urban development, they might prefer the "India News Headlines Dataset". If one\'s interest lies in education, the "Education in India" dataset would be most relevant, and if one is interested in agriculture, the "Agriculture Crop Production In India" dataset would be most important.'

The system does not provide a ranking or importance level for these datasets. The importance of a dataset can depend on the specific needs or interests of the user. For example, if a user is interested in public health, they might find the "COVID-19 in India" dataset most important. If a user is interested in news or urban development, they might prefer the "India News Headlines Dataset". If one's interest lies in education, the "Education in India" dataset would be most relevant, and if one is interested in agriculture, the "Agriculture Crop Production In India" dataset would be most important.

In [17]:
question = "Can you summarise these datasets for me ?"
result = qa({"question": question})
output = result['answer']
output

"Sure, here's a summary of the datasets:\n\n1. Dataset-Nr. 1806, Allegro Articles Summarization Dataset: This dataset was created by The Devastator and last updated on 05/12/2023. It contains summarized Allegro articles and is a source-target dataset. Its size is 117MB and it has a usability rating of 1.00. The dataset is available under the CC0: Public Domain license. The topics covered in this dataset include earth and nature, computer science, software, data cleaning, and data visualization.\n\n2. Dataset-Nr. 927, 2017 Kaggle Machine Learning & Data Science Survey: This dataset is a survey on machine learning and data science, created by Mark McDonald and last updated on 27/10/2017. The size of this dataset is 4MB and it has a usability rating of 0.82. This dataset is available under an Open Database license, with the contents copyrighted by the original authors. Topics covered in this dataset include earth and nature, employment, business, education, artificial intelligence, and co

Sure, here's a summary of the datasets:

1. Dataset-Nr. 1806, Allegro Articles Summarization Dataset: This dataset was created by The Devastator and last updated on 05/12/2023. It contains summarized Allegro articles and is a source-target dataset. Its size is 117MB and it has a usability rating of 1.00. The dataset is available under the CC0: Public Domain license. The topics covered in this dataset include earth and nature, computer science, software, data cleaning, and data visualization.

2. Dataset-Nr. 927, 2017 Kaggle Machine Learning & Data Science Survey: This dataset is a survey on machine learning and data science, created by Mark McDonald and last updated on 27/10/2017. The size of this dataset is 4MB and it has a usability rating of 0.82. This dataset is available under an Open Database license, with the contents copyrighted by the original authors. Topics covered in this dataset include earth and nature, employment, business, education, artificial intelligence, and computer science.

3. Dataset-Nr. 1000, Open Time Series Dataset: This dataset is a collection of time series data for analysis, created by Krishnan S G and last updated on 02/07/2020. Its size is 852KB and it has a usability rating of 0.85. It is available under an Open Database license. The dataset includes topics such as earth and nature, weather and climate, and time series analysis.

4. Dataset-Nr. 153, Kaggle Data Science Survey 2017-2021: This dataset is a compilation of Kaggle data science surveys from 2017 to 2021, created by Andrada and last updated on 26/11/2021. The size of this dataset is 18MB and it has a usability rating of 0.94. It is available under a license specified in the description. Topics covered in this dataset include computer science, exploratory data analysis, survey analysis, and tabular data.

In [18]:
question = "Can you give me the datasets about Germany?"
result = qa({"question": question})
output = result['answer']
output

'Sure, here are the datasets related to Germany:\n\n1. Dataset-Nr.: 1928 \n   Dataset_name: tompaulat/un-sustainable-development-goals-in-germany \n   Title: Sustainable Development Goals \n   Description: Sustainable Development Goals by Tom Paulat. Has Germany achieved its Goals on Sustainability? \n   Last updated: 12/10/2022 \n   Size: 275KB \n   Usability rating: 0.91 \n   Available under CC0: Public Domain license \n   Find more at: https://www.kaggle.com/datasets/tompaulat/un-sustainable-development-goals-in-germany \n\n2. Dataset-Nr.: 31 \n   Dataset_name: abhishek/10k-german-news-articles \n   Title: 10k German News Articles \n   Description: 10k German News Articles by Abhishek Thakur. A dataset for topic extraction and NLP for German language \n   Last updated: 07/11/2019 \n   Size: 123MB \n   Usability rating: 0.82 \n   Available under CC BY-NC-SA 4.0 license \n   Find more at: https://www.kaggle.com/datasets/abhishek/10k-german-news-articles \n\n3. Dataset-Nr.: 862 \n   Da

Sure, here are the datasets related to Germany:

1. Dataset-Nr.: 1928
   Dataset_name: tompaulat/un-sustainable-development-goals-in-germany
   Title: Sustainable Development Goals
   Description: Sustainable Development Goals by Tom Paulat. Has Germany achieved its Goals on Sustainability?
   Last updated: 12/10/2022
   Size: 275KB
   Usability rating: 0.91
   Available under CC0: Public Domain license
   Find more at: https://www.kaggle.com/datasets/tompaulat/un-sustainable-development-goals-in-germany

2. Dataset-Nr.: 31
   Dataset_name: abhishek/10k-german-news-articles
   Title: 10k German News Articles
   Description: 10k German News Articles by Abhishek Thakur. A dataset for topic extraction and NLP for German language
   Last updated: 07/11/2019
   Size: 123MB
   Usability rating: 0.82
   Available under CC BY-NC-SA 4.0 license
   Find more at: https://www.kaggle.com/datasets/abhishek/10k-german-news-articles

3. Dataset-Nr.: 862
   Dataset_name: jihyeseo/german-literature-from-digbiborg
   Title: German Literature from DigBib.Org
   Description: German Literature from DigBib.Org by Jihye Sofia Seo. Literary works in German (poem novels etc)
   Last updated: 20/04/2018
   Size: 8MB
   Usability rating: 0.75
   Available under CC0: Public Domain license
   Find more at: https://www.kaggle.com/datasets/jihyeseo/german-literature-from-digbiborg

4. Dataset-Nr.: 1845
   Dataset_name: thedevastator/mammal-abundance-and-behavior-in-response-to-hun
   Title: Mammal Abundance and Behavior in Response to
   Description: Mammal Abundance and Behavior in Response to by The Devastator. Comparative Study of Germany and United States
   Last updated: 18/02/2023
   Size: 336KB
   Usability rating: 0.94
   Available under CC0: Public Domain license
   Find more at: https://www.kaggle.com/datasets/thedevastator/mammal-abundance-and-behavior-in-response-to-hun

In [19]:
question = "But I want 15 datasets?"
result = qa({"question": question})
output = result['answer']
output

'I\'m sorry, but I can only provide information on the four datasets related to Germany that are mentioned in the context: \n\n1. "10k German News Articles" - a dataset for topic extraction and NLP for the German language. \n\n2. "German Literature from DigBib.Org" - a dataset containing literary works in German. \n\n3. "Sustainable Development Goals" - a dataset investigating whether Germany has achieved its goals on sustainability.\n\n4. "The Big Five European soccer leagues data" - a dataset containing data from the Big Five European soccer leagues, which includes Germany\'s Bundesliga.\n\nFor more datasets related to Germany, you might want to explore the Kaggle datasets page.'

I'm sorry, but I can only provide information on the four datasets related to Germany that are mentioned in the context:

1. "10k German News Articles" - a dataset for topic extraction and NLP for the German language.

2. "German Literature from DigBib.Org" - a dataset containing literary works in German.

3. "Sustainable Development Goals" - a dataset investigating whether Germany has achieved its goals on sustainability.

4. "The Big Five European soccer leagues data" - a dataset containing data from the Big Five European soccer leagues, which includes Germany's Bundesliga.

For more datasets related to Germany, you might want to explore the Kaggle datasets page.

Let's see what's happening under the hood.

In [22]:
qa

ConversationalRetrievalChain(memory=ConversationBufferMemory(chat_memory=ChatMessageHistory(messages=[HumanMessage(content='Give me the datasets about India?'), AIMessage(content='1. Dataset Name: India News Headlines Dataset\n   Dataset Number: 1904\n   Description: This dataset by Rohit Kulkarni contains twenty one years of headlines focusing on India. \n   Size: 93MB\n   Usability rating: 1.00\n   Tags: cities and urban areas, india, business, nlp, text, news\n   More info: [Link here](https://www.kaggle.com/datasets/therohk/india-headlines-news-dataset)\n   \n2. Dataset Name: Hotels in India\n   Dataset Number: 11\n   Description: This dataset by Aakash Shinde contains details about hotels in India such as place, travel sustainability, and ratings.\n   Size: 223KB\n   Usability rating: 0.82\n   Tags: travel, hotels and accommodations\n   More info: [Link here](https://www.kaggle.com/datasets/aakashshinde1507/hotels-in-india)\n   \n3. Dataset Name: COVID-19 in India\n   Dataset Numb

In [29]:
qa.memory.chat_memory.messages

[HumanMessage(content='Give me the datasets about India?'),
 AIMessage(content='1. Dataset Name: India News Headlines Dataset\n   Dataset Number: 1904\n   Description: This dataset by Rohit Kulkarni contains twenty one years of headlines focusing on India. \n   Size: 93MB\n   Usability rating: 1.00\n   Tags: cities and urban areas, india, business, nlp, text, news\n   More info: [Link here](https://www.kaggle.com/datasets/therohk/india-headlines-news-dataset)\n   \n2. Dataset Name: Hotels in India\n   Dataset Number: 11\n   Description: This dataset by Aakash Shinde contains details about hotels in India such as place, travel sustainability, and ratings.\n   Size: 223KB\n   Usability rating: 0.82\n   Tags: travel, hotels and accommodations\n   More info: [Link here](https://www.kaggle.com/datasets/aakashshinde1507/hotels-in-india)\n   \n3. Dataset Name: COVID-19 in India\n   Dataset Number: 1735\n   Description: This dataset by SRK provides data on the Novel Corona Virus Disease 2019 i

In [32]:
qa.memory.chat_memory.messages[0].content

'Give me the datasets about India?'

In [33]:
for i in qa.memory.chat_memory.messages:
  print(i.content)
  print("-"*200)

Give me the datasets about India?
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1. Dataset Name: India News Headlines Dataset
   Dataset Number: 1904
   Description: This dataset by Rohit Kulkarni contains twenty one years of headlines focusing on India. 
   Size: 93MB
   Usability rating: 1.00
   Tags: cities and urban areas, india, business, nlp, text, news
   More info: [Link here](https://www.kaggle.com/datasets/therohk/india-headlines-news-dataset)
   
2. Dataset Name: Hotels in India
   Dataset Number: 11
   Description: This dataset by Aakash Shinde contains details about hotels in India such as place, travel sustainability, and ratings.
   Size: 223KB
   Usability rating: 0.82
   Tags: travel, hotels and accommodations
   More info: [Link here](https://www.kaggle.com/datasets/aakashshinde1507/hotels-in-india)
   
3. Dataset N